### PR TITLE
[#27] Generate the "sample" project from the template

### LINKED ISSUE #27: Add a workflow to generate the "sample" project
## Why

To verify the generated project by the template, we should have a workflow to generate the `sample` project and commit changes to the repository **on each PR**. That means, after making and pushing some changes in the template structure into a PR, the new workflow that will generate and push a new commit directly into that PR to show the new updates for the sample project.

This approach has some concerns are:
- Pushing a new commit via a bot account after a commit from the author will cause the author can not push new commits without pulling or force-pushing 👉 Since, the bot pushes the new commits to update the project, `the author can ignore those and force-push new update commits into their PRs`. The bot will push new commits afterward by combining all the changes in the final sample project ✅ 
- The bot will `not push an update commit without changes` ✅ 
- We can review both all changes from `the template structure` and `the generated project result` (sample) at the same time in one PR ✅ 

> [!NOTE]  
> We must build a simple initial script to generate the template to push the changes.

## Who Benefits?

Developers

### PR DESCRIPTION
**Pull request by luongvo**
- Close #27

## What happened 👀

- Create a simple `make.sh` to generate the project from the template, which basically copies all project files to the `sample` folder, except `.git` and `sample`.
- Create a test workflow with a job "generate_sample_project" to generate the project.

## Insight 📝

- We need to check out the source code with the corresponding working `branch name` for pushing up. Otherwise, it will throw the error `fatal: You are not currently on a branch.` 💥  https://github.com/nimblehq/kmm-templates/actions/runs/5862694160/job/15894905375.
  - To check out to the corresponding working `branch name`, use `ref: ${{ github.head_ref }}` https://laurentsenta.com/digital-garden/programming/github-actions-tips/ ✅ 
- The job `generate_sample_project` to generate the project will be executed after the job `test` is completed successfully.

## Proof Of Work 📹

- There is a new commit to push the `sample` project changes on each PR for verifying https://github.com/nimblehq/kmm-templates/pull/28/commits

image

👉 A new `sample` project folder for reference https://github.com/nimblehq/kmm-templates/tree/feature/27-generate-sample-project/sample

### CODE REVIEW

File: .github/workflows/test.yml
Comment (luongvo) [inline .github/workflows/test.yml:1]: FYI: there are only 2 file changes from this PR: this one and `make.sh`.

File: sample/.github/CODEOWNERS [hunk 1]
```diff
@@ -0,0 +1,3 @@
+# Team
+# @luongvo is the Team Lead
```
Comment (lydiasama) [inline sample/.github/CODEOWNERS]:
```suggestion
# @phongvhd93 is the Team Lead
```
😛 😛

File: sample/.gitmodules
Comment (luongvo) [inline sample/.gitmodules:1]: exclude this from the generated project

File: sample/ios/.github/CODEOWNERS
Comment (luongvo) [inline sample/ios/.github/CODEOWNERS:1]: skip this `ios/.github` folder

File: sample/.github/workflows/test.yml [hunk 2]
```diff
@@ -0,0 +1,50 @@
+name: Test
+on:
+  pull_request:
+    types: [ opened, synchronize, reopened ]
+
```
Comment (lydiasama) [inline sample/.github/workflows/test.yml:5]: Shall we put the same thing as in templates?
```suggestion
  push:
    branches-ignore:
      - develop
      - 'release/**'
      - main
```

Or I guess you will rebase later? 🤔

File: sample/ios/README.md
Comment (luongvo) [inline sample/ios/README.md:1]: skip

File: sample/ios/bitrise.yml
Comment (luongvo) [inline sample/ios/bitrise.yml:1]: skip

File: sample/ios/codemagic.yaml
Comment (luongvo) [inline sample/ios/codemagic.yaml:1]: skip

File: sample/ios/deliverable_setup.sh
Comment (luongvo) [inline sample/ios/deliverable_setup.sh:1]: skip

File: sample/.github/workflows/test.yml [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        uses: actions/checkout@v3
+        with:
+          ref: ${{ github.head_ref }}
+          submodules: 'recursive'
+
+      - name: Remove the old sample project
+        run: rm -rf sample
+
+      - name: Generate the new sample project
+        run: ./make.sh --bundle-id co.nimble.kmm.sample --bundle-id-staging co.nimble.kmm.sample.staging --project-name sample --ios-version 14.0
```
Comment (lydiasama) [inline sample/.github/workflows/test.yml:35]:
```suggestion
        run: ./make.sh --bundle-id co.nimble.kmm.sample --bundle-id-staging co.nimble.kmm.sample.staging --project-name sample --ios-version ${IOS_VERSION}
```
Can we put this to be the variable?

File: sample/make.sh
Comment (luongvo) [inline sample/make.sh:1]: skip

File: README.md [hunk 4]
```diff
@@ -2,7 +2,13 @@
 
 Kotlin Multiplatform project repository template to set up all projects at [Nimble](https://nimblehq.co/)
```
Comment (lydiasama) [inline README.md]:
```suggestion
Kotlin Multi-platform Mobile (KMM) template to set up all KMM projects at [Nimble](https://nimblehq.co/)
```

File: .github/workflows/test.yml [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+        uses: actions/checkout@v3
+
+      # TODO: More CI steps will be added in https://github.com/nimblehq/kmm-templates/issues/4
+
+  generate_sample_project:
+    # The job "generate_sample_project" identify that the job "test" must complete successfully
+    # before this "generate_sample_project" job will run
+    needs: test
+    name: Generate the sample project
+    runs-on: macOS-12
```
Comment (suho) [inline .github/workflows/test.yml]:
```suggestion
    runs-on: macos-latest
```

File: README.md (hunk 4 above)
Comment (luongvo) [reply README.md]: @lydiasama Multiplatform is a single word https://kotlinlang.org/docs/multiplatform.html

```suggestion
Kotlin Multiplatform Mobile (KMM) template to set up all KMM projects at [Nimble](https://nimblehq.co/)
```

File: sample/.github/workflows/test.yml (hunk 3 above)
Comment (luongvo) [reply sample/.github/workflows/test.yml:35]: @lydiasama I raised my concern on this PR https://github.com/nimblehq/kmm-templates/pull/24/files#r1295466511. Let @phongvhd93 handle this later 🙏

File: README.md
Comment (luongvo) [inline README.md:1]: @phongvhd93 FYI: I added `Use the template` section for the template in this PR

File: sample/.github/workflows/test.yml (hunk 2 above)
Comment (luongvo) [reply sample/.github/workflows/test.yml:5]: @lydiasama We already have `pull_request` to verify the template script. Add `push` event is redundant here and will cause conflict between these events when pushing sample project changes.
- push:  https://github.com/nimblehq/kmm-templates/actions/runs/5875799533/job/15932720203 ✅ 
- pull_request: https://github.com/nimblehq/kmm-templates/actions/runs/5875800008 ❌

Comment (luongvo) [reply sample/.github/workflows/test.yml:5]: One `pull_request` event is enough https://github.com/nimblehq/kmm-templates/actions/runs/5876269164?pr=28 ✅

File: README.md [hunk 6]
```diff
@@ -11,19 +17,21 @@ Kotlin Multiplatform project repository template to set up all projects at [Nimb
 
 Please check [this](https://kotlinlang.org/docs/multiplatform-mobile-setup.html#install-the-necessary-tools) out for more details.
```
Comment (phongvhd93) [inline README.md:17]: We should also mention about:
```
Ruby 3.1.2
Xcode 13.3+
```

Comment (luongvo) [reply README.md:17]: @phongvhd93 we need Ruby to generate the iOS project or to run the final generated KMM project?

Comment (luongvo) [reply README.md:17]: @phongvhd93 I think we don't need to mention 
- Ruby, I can even run the script with my Ruby 2.6.10 🤷‍♂️ macOS default env has Ruby already.
- [Tuist](https://github.com/tuist/tuist) as the script will try to grab the missing for us.

Comment (luongvo) [reply README.md:17]: Updated in https://github.com/nimblehq/kmm-templates/pull/28/commits/ebc9518f2585cbd63933973fc0ba6f11c769b7ca

**Review by phongvhd93 — APPROVED** [review]

**Comment by luongvo** [comment]
@suho @lydiasama @AVI5HEK @markgravity @kaungkhantsoe We need 1 more approval on this PR to process the next steps 👋

File: README.md [hunk 7]
```diff
@@ -1,29 +1,36 @@
...
-TODO
+    `$ git submodule update --init --recursive`
 
-### Test
+3. Generate the new project by running the following command.
 
-TODO
+    `$ ./make.sh --bundle-id co.nimble.kmm.sample --bundle-id-staging co.nimble.kmm.sample.staging --project-name sample --ios-version 14.0`
+
+The generated project is at `/sample` folder and ready to build, run and test 🎉
```
Comment (AVI5HEK) [inline README.md]:
```suggestion
The generated project is at `/sample` folder and ready to build, run, and test 🎉
```

**Review by AVI5HEK — APPROVED** [review]
The rest LGTM. Approving nevertheless.

**Pull request merged**